pkg/service/healthcheck/service_integration_test.go

```diff
@@ -547,6 +553,36 @@ func tryStartAgent(t *testing.T, hosts []string) {
 	}
 }
 
+func ensureNodesAreUP(t *testing.T, hosts []string, timeout time.Duration) error {
```

Comment:
Can't we just create a client with all nodes and get their status from there?
If it's important to get how each nodes self reports its status, instead of the statuses perceived by a single node, we could either create per node client or `forceHost` in status requests.
It would allow us to skip manual output parsing and make things easier in general.

Reply:
We could use the client.
But if someone called to blockScyllaAPI before, then it will never return the status found with the gossiper as the call will be blocked.
Let's leave the "nodetool status" to do this check.

Reply:
> But if someone called to blockScyllaAPI before, then it will never return the status found with the gossiper as the call will be blocked.

It should return after timeout, right? If we want for it to be quick, we can set the timeout to a low value.

Reply:
It's not about the timeout only, but about actual response.
If the API is blocked I won't know which nodes are UP/DN, but I do know it when I use nodetool status.

Reply:
Okay, but getting back to the beginning - we have a test that stops a node and blocks its API. At the end of this test we want to start the node and unblock its API. We also want to wait for those changes to really set in, so that they don't affect other tests by accident.

1. Why can't we simply create a client to all of the nodes and send some per host requests (e.g. `client.Snapshots`) to all of them sequentially? Node needs to be both up and responsive on API for it to succeed. We can do it sequentially, because it's not like querying the endpoint does any work - it just check responsiveness, so making it in parallel does not provide visible time improvement, but it adds some complexity to the code. If you prefer to do it in parallel, we should use already implemented and tested functions to do so (e.g. `parallel.Run`)
2. Shouldn't we add this waiting step to the test that stops nodes? I would say that it's easier to to remember to run this function at the end of the test that modifies node state, instead of running it at the beginning of every test that requires nodes to be in the correct state (so basically all of the tests). In this PR you encountered this problem only for a single test, but that's because of given sequential order of test execution that we shouldn't rely on.

Reply:
> Why can't we simply create a client to all of the nodes and send some per host requests (e.g. client.Snapshots) to all of them sequentially? Node needs to be both up and responsive on API for it to succeed. We can do it sequentially, because it's not like querying the endpoint does any work - it just check responsiveness, so making it in parallel does not provide visible time improvement, but it adds some complexity to the code. If you prefer to do it in parallel, we should use already implemented and tested functions to do so (e.g. parallel.Run)

That's an interesting point, but I think there's a distinction here. A node doesn't necessarily need to be fully "UP" to respond to some API calls—it could still be in a "Joining" state, for example.
Using nodetool status gives us a more reliable check to ensure the node is fully operational (UP) as part of the cluster. For this test, I also want to ensure the agent service is running and that ports are unblocked, which nodetool helps verify comprehensively.

> Shouldn't we add this waiting step to the test that stops nodes? I would say that it's easier to to remember to run this function at the end of the test that modifies node state, instead of running it at the beginning of every test that requires nodes to be in the correct state (so basically all of the tests). In this PR you encountered this problem only for a single test, but that's because of given sequential order of test execution that we shouldn't rely on.

That's a valid point, and I agree it could simplify some workflows if we add the waiting step to the test that stops nodes. However, I still prefer to harden and explicitly check the assumptions for this specific test to ensure it behaves predictably.

Reply:
> That's an interesting point, but I think there's a distinction here. A node doesn't necessarily need to be fully "UP" to respond to some API calls—it could still be in a "Joining" state, for example.
Using nodetool status gives us a more reliable check to ensure the node is fully operational (UP) as part of the cluster. For this test, I also want to ensure the agent service is running and that ports are unblocked, which nodetool helps verify comprehensively.

Nodetool is just a simple client that uses Scylla API - very similar to what agent does. Agent can also issue status API call - the only problem here is that our implementation does not allow to specify host (SM kind of assumes that it doesn't matter which host will be chosen). But in order to deal with that, we can simply apply `ClientContextWithSelectedHost` (exported `forceHost`).

I just don't like using shell commands and parsing raw text when we already have a proper way of getting this information in our repo.

Also, the comment about using `parallel.Run` still stands.

> That's a valid point, and I agree it could simplify some workflows if we add the waiting step to the test that stops nodes. However, I still prefer to harden and explicitly check the assumptions for this specific test to ensure it behaves predictably.

I'm not against validating it here, but we could probably save some time in the future if we also validate it after `TestValidateHostConnectivityIntegration`. Otherwise, some newly written test might start failing on CI, when they pass locally, and it could be annoying to debug.

Reply:
@Michal-Leszczynski I'm not gonna change it. Let's leave it parallel and include `nodetool status`.

Reply:
What about the last comment regarding `TestValidateHostConnectivityIntegration`?

Reply:
I want this part with ensureNodesAreUP to help identify the root cause of the situation where a node is down, but it was expected to be up. Without it, the test error may contain misleading information that is just a result of the test assuming an incorrect cluster state.

So, it's just one step. It still doesn't point to the exact root cause.
The root cause lies in a previous test that was incorrectly implemented and didn’t clean up properly after execution. This means the test that failed was actually correct.

In this PR, this is exactly the case. TestStatus_Ping_Independent_From_REST_Integration was failing, but the reason for the failure was the TestValidateHostConnectivityIntegration test introduced in this PR. The former stopped failing after applying a fix to TestValidateHostConnectivityIntegration.

IMHO, it's better to be explicit in TestStatus_Ping_Independent_From_REST_Integration about the reason for the failure. In this case, the issue was that TestValidateHostConnectivityIntegration did not restart the service after execution.

By adding ensureNodesAreUP, it cuts the test execution and explicitly indicates that the Scylla service is not running.

What already existed in TestStatus_Ping_Independent_From_REST_Integration, such as:

```
tryUnblockCQL(t, ManagedClusterHosts())
tryUnblockREST(t, ManagedClusterHosts())
tryUnblockAlternator(t, ManagedClusterHosts())
tryStartAgent(t, ManagedClusterHosts())
```
... was not enough to terminate the execution early, which allowed the test to continue.

By creating a client and calling some random endpoint on it, I could end up in a situation where previous tests cleaned up after themselves, but since supervisorctl start is asynchronous, it was not fast enough to complete before the current test started.

Reply:
Code is updated to make the block/unblock/try methods public and to use them in TestValidateHost... tests as well.